### PR TITLE
test(nn): add unit tests for autoencoder library (#87)

### LINKED ISSUE #87: Tests for `autoencoder` library
Before merging `dev` into main, we should write some basic unit tests for the autoencoder library

### PR DESCRIPTION
**Pull request by lucacirfeta**
Resolves #87

### Summary
This PR adds unit test coverage for the `ml4gw.nn.autoencoder` subpackage, including base abstractions, convolutional autoencoders, skip connection modules, and resizing utilities.

> [!NOTE]
> **Note for reviewers:** While writing the tests, I encountered and fixed 4 runtime bugs in the underlying autoencoder code (detailed under *Small bugfixes* below). Because these modify runtime behavior rather than just adding test files, I wanted to flag them explicitly — in particular the `ConvolutionalAutoencoder` skip-connection adjustment.
> 
> If you prefer to keep this PR strictly test-only and handle the fixes in a dedicated PR (or discuss any architectural preference on the skip connection handling), I am very happy to split them.

### What is covered
- `ml4gw.nn.autoencoder.base.Autoencoder`:
  - Empty, single-block, and multi-block sequential pipelines.
  - Isolated calls to `encode` and `decode` (with and without `return_states`).
  - Validation of intermediate states required by skip connections (raising explicit `ValueError` when missing or mismatched).
- `ml4gw.nn.autoencoder.skip_connection`:
  - Output channel arithmetic (`get_out_channels`) for base, add, and concat skip connections.
  - Forward dimension matching and channel interleaving for grouped `ConcatSkipConnect`.
  - Explicit error handling when input channels cannot be divided evenly across groups.
- `ml4gw.nn.autoencoder.convolutional`:
  - `ConvBlock` and `ConvolutionalAutoencoder` parametrized across channels, kernel sizes (3, 5), strides (1, 2), groups, and custom output activations.
  - Exact output shape preservation across arbitrary input lengths via `match_size`.
  - Gradient backward pass test (`loss.backward()`) ensuring all trainable parameters receive finite, non-NaN gradients.
- `ml4gw.nn.autoencoder.utils.match_size`:
  - Unit tests covering exact match, zero-padding (odd/even), and symmetric cropping (odd/even).

### Small bugfixes revealed during testing
- Fixed slicing in `match_size` when `target_size < X.size(-1)`.
- Fixed intermediate state indexing in `Autoencoder.decode` (`states[i]` instead of double reverse index).
- Ensured uninstantiated `activation` classes passed to `ConvBlock` are initialized as callable modules.
- Adjusted skip connection attachment in `ConvolutionalAutoencoder` so the bottleneck layer decodes directly without expecting an upstream skip state.

### What is explicitly NOT covered
- 2D / 3D spatial autoencoders (currently only 1D conv architectures are defined in this module).
- Dynamic argument inspection between skip connection and autoencoder blocks (marked as TODO in source).

### Verification
Ran locally with `pytest`:
- `tests/nn/autoencoder/`: 166 passed.
- Full test suite: 1948 passed.
- Pre-commit formatting (`uv run prek run --all-files`): all checks passed.

### CODE REVIEW

**Comment by wbenoit26** [comment]
Hi @lucacirfeta, thanks for this PR and thanks for your patience as we've had people traveling/on vacation. I'll try to review this in the next day or two, though from a glance it looks like a useful addition.

File: tests/nn/autoencoder/test_autoencoder_utils.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import pytest
+import torch
+
+from ml4gw.nn.autoencoder.utils import match_size
+
+
+@pytest.mark.parametrize("target_size", [8, 16, 23, 32])
+@pytest.mark.parametrize("input_size", [8, 15, 16, 27, 32])
+def test_match_size(target_size, input_size):
```
Comment (wbenoit26) [inline tests/nn/autoencoder/test_autoencoder_utils.py:9]: This test re-implements the function itself, so it doesn't serve to check whether the function does what we want. Could you instead have the test locate the content of the smaller tensor within the larger tensor and test that `left` and `right` are within 1 sample of each other?

E.g., if the tensor is getting expanded, locate the original tensor within the padded version, measure its distance from each edge, and confirm that `abs(left - right) <= 1`.

File: tests/nn/autoencoder/test_base.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    # Forward
+    out = ae(x)
+    assert out.shape == x.shape
+
+
+def test_autoencoder_with_skip_connection():
+    skip = AddSkipConnect()
+    ae = Autoencoder(skip_connection=skip)
+    # Using blocks with equal in/out channels
+    # so AddSkipConnect can add x + state
```
Comment (wbenoit26) [inline tests/nn/autoencoder/test_base.py]: I suspect that the reason that the input/output channels need to be the same is the indexing bug that you fixed in `base.py`. With that fix in place, can you try uneven channels?

File: tests/nn/autoencoder/test_convolutional.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+)
+def test_convolutional_autoencoder_shapes(
+    in_channels,
+    encode_channels,
+    kernel_size,
+    stride,
+    decode_channels,
+    skip_connection,
+):
+    # When using AddSkipConnect with ConvolutionalAutoencoder,
```
Comment (wbenoit26) [inline tests/nn/autoencoder/test_convolutional.py]: Likewise here. I think these lines can be removed.

File: ml4gw/nn/autoencoder/convolutional.py
Comment (wbenoit26) [inline ml4gw/nn/autoencoder/convolutional.py]: We can simplify the changes below with the type definition from line 10. In the signature, we can set:
```python
activation: Module = torch.nn.ReLU,
...
output_activation: Module | None = None,
```

And then in the body:
```python
self.activation = activation()
self.output_activation = (output_activation or activation)()
```

All the checks can then be removed.

File: ml4gw/nn/autoencoder/base.py
Comment (wbenoit26) [inline ml4gw/nn/autoencoder/base.py]: Pre-existing issue that I'm noticing on this review: we don't raise an error when `self.skip_connection` is `None`, but `states` is not `None`. Would you mind adding a check for that?

**Review by wbenoit26 — CHANGES_REQUESTED** [review]
This looks great – thanks for adding tests and making the autoencoder module usable. There's a few things that I'd like to see changed in the code; I've commented on them below.

A couple other items that would be nice to have, but that I wouldn't hold up this PR over:
1. Benchmarks: In `benchmarks/nn` we have benchmarks against resnet, s4d, etc, but none for the autoencoder. Would be great to add some here as well.
2. Docstrings: Now that the code is functional, docstrings for `ConvBlock` and `match_size` would be very welcome.

Also, we [recently added](https://github.com/ML4GW/ml4gw/pull/339) an AI disclosure policy. I'm not concerned about the content of this PR, but for completeness, if you used AI, could you please add a statement?

**Comment by lucacirfeta** [comment]
Thanks for review!

I addressed the requested changes:

- Reworked the `match_size` test to locate the original content and verify
  that the left/right padding or cropping differs by at most one sample.
- Tested skip connections with uneven channel sizes.
- Removed the channel-equality workaround from the convolutional tests.
- Simplified activation handling to use module constructors.
- Added an explicit error when intermediate states are passed without a
  configured skip connection.
- Added docstrings for `ConvBlock` and `match_size`.

Validation:
- `459 passed` in `tests/nn`
- all relevant `prek` hooks passed

### AI disclosure

AI tools were used to assist with review analysis, implementation, and test
design. I reviewed the final changes, verified the resulting diff, and ran
the tests reported in this PR.

**Review by wbenoit26 — APPROVED** [review]

**Comment by wbenoit26** [comment]
Great, thanks for making these updates.
